src/art/skypilot/api.py

```diff
@@ -0,0 +1,203 @@
+from sky.core import endpoints
+import os
+import semver
+
+from art.skypilot.load_env_file import load_env_file
+from .utils import is_task_created, wait_for_task_to_start, to_thread_typed
+
+from .. import dev
+from ..api import API
+
+if TYPE_CHECKING:
+    from ..model import Model, TrainableModel
+
+
+class SkyPilotAPI(API):
+    _cluster_name: str
+
+    def __init__(
+        self,
+        *,
+        cluster_name: str = "art",
+    ) -> None:
+        self._cluster_name = cluster_name
+
+    async def initialize_cluster(
+        self,
+        *,
+        resources: sky.Resources | None = None,
+        art_version: str | None = None,
+        env_path: str | None = None,
+    ) -> None:
+        if resources is None:
+            resources = sky.Resources(
+                cloud=sky.clouds.RunPod(),
+                # region="US",
+                accelerators={"H100": 1},
+                ports=[],
+            )
+
+        # ensure ports 7999 and 8000 are open
+        updated_ports = resources.ports
+        if updated_ports is None:
+            updated_ports = []
+        updated_ports += ["7999", "8000"]
+        resources = resources.copy(ports=updated_ports)
+
+        # check if cluster already exists
+        cluster_status = await to_thread_typed(
+            lambda: sky.status(cluster_names=[self._cluster_name])
+        )
+        if (
+            len(cluster_status) == 0
+            or cluster_status[0]["status"] != sky.ClusterStatus.UP
+        ):
+            await self._launch_cluster(resources, art_version, env_path)
+        else:
+            print(f"Cluster {self._cluster_name} exists, using it...")
+
+        if await is_task_created(
+            cluster_name=self._cluster_name, task_name="art_server"
+        ):
+            print("Art server task already running, using it...")
+        else:
+            art_server_task = sky.Task(name="art_server", run="uv run art")
+            resources = await to_thread_typed(
+                lambda: sky.status(cluster_names=[self._cluster_name])[0][
+                    "handle"
+                ].launched_resources
+            )
+            art_server_task.set_resources(resources)
+
+            # run art server task
+            await to_thread_typed(
+                lambda: sky.exec(
+                    task=art_server_task,
+                    cluster_name=self._cluster_name,
+                    detach_run=True,
+                )
+            )
+            print("Task launched, waiting for it to start...")
+            await wait_for_task_to_start(
+                cluster_name=self._cluster_name, task_name="art_server"
+            )
+            print("Art server task started")
+
+        art_endpoint = await to_thread_typed(
+            lambda: endpoints(cluster=self._cluster_name, port=7999)[7999]
+        )
+        base_url = f"http://{art_endpoint}"
+        print(f"Using base_url: {base_url}")
+
+        super().__init__(base_url=base_url)
```

Comment:
Might be good to add a check here that we can actually connect to the remote port as expected. And maybe put it in a little loop that checks once a sec for 10 secs or something in case it takes a few seconds for the remote server to actually start listening.

Reply:
That's a good idea. Are you imagining a `/healthcheck` endpoint on the server to ping?

Reply:
Yep, that's a good way to do it.